src/ejabberd_c2s.erl

```diff
@@ -491,7 +512,12 @@ wait_for_feature_before_auth({xmlstreamelement, El}, StateData) ->
         {?NS_SASL, <<"auth">>} when TLSEnabled or not TLSRequired ->
             Mech = xml:get_attr_s(<<"mechanism">>, Attrs),
             ClientIn = jlib:decode_base64(xml:get_cdata(Els)),
-            StepResult = cyrsasl:server_start(StateData#state.sasl_state, Mech, ClientIn),
+            SaslState = StateData#state.sasl_state,
+            Server = StateData#state.server,
+            AuthMech = [M || M <- cyrsasl:listmech(Server), filter_mechanism(M, StateData)],
```

Comment:
I'd rename it to something that says these are the advertised mechanisms to this user, or at least that it is a plural thing or so. Same for the related key in the map created below. Also, this was calculated already in the previous step of the fsm (look for the previous call to `cyrsasl:listmech/1`), would be cool if we could keep it and not having to calculate it again. That's something I might profile some day and see it's relative impact, remind me to do so when I have time 😄

Reply:
Yes, the name could be more accurate here. This line is already quite long and I didn't wanted to split it to multiple lines only because of the variable length. I was having hard time finding here something short yet descriptive. 
listmech is already done before but it's wrapped quite extensively in XML as it's return of define_features. There could be 3 ways of approaching this:
* Storing full return value of define features in c2s state and unwrap later when checking what features were advertised. Not sure if this would make things faster than extracting the values twice. Also the c2s state will grow yet again.
* Returning just the simple format from define_features, storing them in c2s and wrapping in XML later, although the mechanisms are not the only features that we advertise and it's quite nested. So we would nest in to get the features and then nest in again to create proper XML. Here again we'd need to add the info to c2s state.
* Doing this one line of code is getting simple return value simple and not adding anything new to c2s state.

Reply:
I'm against extending c2s state at this point. When we find this redundant filtering is a bottleneck, then we can think about optimisation.